Fix incorrect results of aggregate views withoug GROUP BY

yugo-n opened the pull request:
Views of aggregate without GROUP BY clause always have
only one tuple as the result. Even when all tuples in base
tables are deleted, the view must have a row whose result is
0 for count(), or NULL for others.

However, in the previous implementation, this resulted in
deleting the tuple from the view.  To fix this, don't allow
to delete a tuple from a view when this is a aggregate view
without GROUP BY.

Github issue #88

yugo-n committed:
Fix incorrect results of aggregate views withoug GROUP BY

Views of aggregate without GROUP BY clause always have
only one tuple as the result. Even when all tuples in base
tables are deleted, the view must have a row whose result is
0 for count(), or NULL for others.

However, in the previous implementation, this resulted in
deleting the tuple from the view.  To fix this, don't allow
to delete a tuple from a view when this is a aggregate view
without GROUP BY.

Github issue #88

thoshiai reviewed (commented):
this work is fine.
However, I'm concerned about 'tuptable_recalc' variable.
Do you need this variable? I think it is not necessary.

yugo-n commented:
> this work is fine.
> However, I'm concerned about 'tuptable_recalc' variable.
> Do you need this variable? I think it is not necessary.

Well, do you mean that we can use SPI_tuptable directly, instead of tuptable_recalc argument,  in order to pass the query result from apply_old_delta() to its caller?  As for it, you are right, this is not necessary, but we can conceal the SPI interface which uses the global variable via this function argument, and this is what I intend.

thoshiai commented on src/backend/commands/matview.c:
```diff
@@ -3451,24 +3453,39 @@ build_query_for_apply_old_delta(const char *matviewname, const char *deltaname_o
+
+	if (minmax_list)
+	{
+		if (tuptable_recalc)
```
I think that Assert() is better than this if statement.
tuptable_recalc must be true.

yugo-n replied:
This makes sense. Because apply_old_delta is a static function so we can easily assume this argument is not null. I'd fixed it.

thoshiai reviewed (commented):
Thank you for your reply. I understand your thinking, it is good.

I add a suggestion of small fix.

yugo-n committed:
Add an assert macro that ensures tuptable_recalc is not null

thoshiai commented on src/backend/commands/matview.c:
```diff
@@ -3451,24 +3453,39 @@ build_query_for_apply_old_delta(const char *matviewname, const char *deltaname_o
+	{
+		if (tuptable_recalc)
+			*tuptable_recalc = NULL;
+		return 0;
```
This is my humble opinion, I think it's fine to always return SPI_processed. The return value is simpler.

yugo-n replied:
Hmm, I'd like consider the consistency of returned results more than that simplicity. However, the relation between the return value and tuptable_recalc might have been unclear, so I'd changed it to return the number of rows via a pointer argument.

yugo-n committed:
Fix apply_old_delta to return the number of tuples to be recalculated via argument

This seems better interface because the work to return information
of tuples to be ecalclulated is not the primary objection of this
function.

thoshiai commented:
Thank you for your fix. I think it is no problem.
I will merge it.
